sphinx/ext/apidoc/_extension.py

```diff
@@ -0,0 +1,231 @@
+"""Sphinx extension for auto-generating API documentation."""
+
+from __future__ import annotations
+
+import fnmatch
+import os
+import re
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any
+
+from sphinx.locale import __
+from sphinx.util.console import bold
+
+from . import WARNING_TYPE, _remove_old_files, create_modules_toc_file, logger, recurse_tree
+
+if TYPE_CHECKING:
+    from sphinx.application import Sphinx
+
+
+def run_apidoc_ext(app: Sphinx) -> None:
+    """Run the apidoc extension."""
+    logger.info(bold(__('Running apidoc')))  # TODO use iterater
+
+    options: dict[str, Any]
+    for i, options in enumerate(app.config.apidoc_modules):
+        if not isinstance(options, dict):
+            logger.warning(__('apidoc_modules item %i must be a dict'), i, type=WARNING_TYPE)
+            continue
+
+        # module path should be absolute or relative to the conf directory
+        # TODO account for Windows path?
+        if not (path := _check_string(i, options, 'path', True)):
+            continue
+        module_path = app.confdir.joinpath(path)
+        if not module_path.is_dir():
+            logger.warning(
+                __("apidoc_modules item %i 'path' is not an existing folder: %s"),
+                i,
+                module_path,
+                type=WARNING_TYPE,
+            )
+            continue
+
+        # destination path should be relative to the source directory
+        # TODO account for Windows path?
+        if not (destination := _check_string(i, options, 'destination', True)):
```

Comment:
Lone Booleans at call sites are often hard to interpret -- perhaps make ``required`` a keyword-only argument?